Use Cython to boost performance of metrics

jacobmontiel opened the pull request:
Performance metrics calculation is an expensive operation when performed on single samples instead of batches. A known bottleneck is the extensive usage of `for` loops.

This PR focuses on optimizing parts of the metrics calculations by using Cython implementations.

## Changes required

- [x] Implement a confusion matrix that supports both insertion and deletion (binary and multi-class)
- [x] Implement muti-label confusion matrix
- [ ] ~Refactor the "classification report" so a single instance of the confusion matrix is updated incrementally and used to calculate multiple metrics on demand.~
- [x] Replace python implementation where possible

## Metrics

- [x] accuracy
- [x] kappa
- [x] kappa-t
- [x] kappa-m
- [x] precision (micro, macro, weighted, example-based)
- [x] recall (micro, macro, weighted, example-based)
- [x] f1 (micro, macro, weighted, example-based)
- [x] fbeta (micro, macro, weighted, example-based)
- [x] geometric mean
- [x] roc auc
---
- [x] Hamming score 
- [x] exact match
- [x] Jaccard score

jacobmontiel commented:
@MaxHalford @smastelini This is an ongoing effort. So far I focused on:

- Fix installation as the subtree uses a different approach to process Cython files
- Modify the existing Cython implementation of the confusion matrix for binary and multi-class to be closer to the python implementation
  - Support strings
  - Use `dict` of `dict` to access via `[][]`
  - Provide a similar interface including the `revert` method
  - I did not extend `MultiClassMetric` as I am not familiar with it. This is relevant depending on how we intend to use the confusion matrix, especially within the report.
- I also temporarily brought the `ClassificationPerformanceEvaluator` so we have all the code in a single place. The new "classification report" will likely be a combination of it and the existing code.
- If this looks good then I will proceed to do the modifications in the multi-label version.

MaxHalford commented on setup.py:
```diff
@@ -104,6 +112,18 @@
     ],
     ext_modules=cythonize(
         module_list=[
+            setuptools.Extension(
```
Is this necessary? Doesn't the `sources=['**/*.pyx']` underneath include everything?

jacobmontiel replied:
This is required as the files use numpy arrays. In this case, we must indicate the path to `numpy` which is done via the `get_include` method.

Alternatively, we can define the numpy path as a generic option, but I am not sure about the impact (if any) on cython code that does not use `numpy`.

jacobmontiel replied:
I simplified this in a61a9597c360687d911e4a90eaded8f8088b06cd by setting the path to numpy by default for all cython modules.

MaxHalford commented on creme/metrics/_confusion_matrix.pyx:
```diff
@@ -0,0 +1,232 @@
+
+
+cdef class MultiLabelConfusionMatrix:
+    """ Multi-label Confusion Matrix.
```
Extra space.

jacobmontiel replied:
Removed in a61a9597c360687d911e4a90eaded8f8088b06cd

MaxHalford commented on creme/metrics/_confusion_matrix.pyx:
```diff
@@ -0,0 +1,232 @@
+        return table
+
+
+cdef class MultiLabelConfusionMatrix:
```
Just thinking out loud, isn't `MultiLabelConfusionMatrix` essentially a `dict` of `ConfusionMatrix`?

jacobmontiel replied:
It is possible to see it that way. However, many operations can be easily performed by using a 3D array:

```python
sum_tp = data[:,1,1]
```

MaxHalford commented on creme/metrics/_confusion_matrix.pyx:
```diff
@@ -0,0 +1,232 @@
+
+from collections import defaultdict
+
+cdef bint boolean_variable = True
```
I can't this being used anywhere :)

jacobmontiel replied:
It is used here :)

https://github.com/creme-ml/creme/blob/a61a9597c360687d911e4a90eaded8f8088b06cd/creme/metrics/_confusion_matrix.pyx#L168

MaxHalford replied:
Lol sorry I don't understand: `boolean_variable` is the name of the boolean variable that is declared, but I don't see any reference to `boolean_variable` in the rest of the file.

jacobmontiel replied:
My bad, you are right, that was completely unrelated and unused.

MaxHalford commented on creme/metrics/_confusion_matrix.pyx:
```diff
@@ -0,0 +1,232 @@
+
+    def update(self, y_true, y_pred, sample_weight=1.):
+        if sample_weight is None:
+            # Since we ca not set a default value in the signature
```
cannot

jacobmontiel replied:
Fixed typo in a61a9597c360687d911e4a90eaded8f8088b06cd

MaxHalford commented on creme/metrics/_confusion_matrix.pyx:
```diff
@@ -0,0 +1,232 @@
+        self.sum_col = defaultdict(float)
+        self.data = defaultdict(lambda: defaultdict(float))
+
+    def __setitem__(self, entry, double sample_weight):
```
I don't mind having this, but I can't think of a usecase for it.

MaxHalford commented on creme/metrics/_confusion_matrix.pyx:
```diff
@@ -0,0 +1,232 @@
+
+cdef bint boolean_variable = True
+
+cdef class ConfusionMatrix:
```
How about if we move the methods from `ClassificationPerformanceEvaluator` into this class?

jacobmontiel replied:
Definetly, I will explore this option.

smastelini commented on creme/metrics/report.py:
```diff
@@ -16,109 +17,71 @@ class ClassificationReport(base.MultiClassMetric):
+    >>> print(report)
+               Precision   Recall   F1      Support
+    <BLANKLINE>
+       apple       0.000    0.000   0.000       1.0
+      banana       0.000    0.000   0.000       3.0
+        pear       0.000    0.000   0.000       1.0
```
Isn't something odd with those results?

MaxHalford replied:
Yes indeed... I'll look into it!

MaxHalford commented on creme/metrics/_confusion_matrix.pyx:
```diff
@@ -0,0 +1,286 @@
+    @property
+    def majority_class(self):
+        try:
+            return self._class_counter.most_common(1)[0][0]
```
We can also use the `sum_row` attribute, right? That way we don't need to create and update `_class_counter`.

jacobmontiel replied:
That was the original idea. I used the counter to avoid the loop to get the majority class. However, since we are using Cython this shouldn't be a big issue. Now we are using `sum_row`

jacobmontiel commented:
> Wouldn't it be better to put this logic in the `revert` method? Then we wouldn't need to check if `sample_weight` is negative.

> I would return None :)

I am not sure about what these comments refer to

MaxHalford commented:
Sorry @jacobmontiel my comments got lost in the changes :)

jacobmontiel commented:
No problem @MaxHalford 

I addressed the comment regarding the negative `sample_weight`.

Regarding the usage of `NoChangeClassifier` and `PriorClassifier` seems an overhead as we still need to track the observed weight. The code is a pseudo-classifier only implementing the core elements with the information already available in the confusion matrix. But I am biased toward the current approach. What I like in the current implementation is that any update to the confusion matrix will keep statistics updated for all related metrics. We can also open a follow-up PR to investigate how to improve in this aspect.
